pkg/localstore/reserve.go

```diff
@@ -105,22 +73,91 @@ func (db *DB) UnreserveBatch(id []byte, radius uint8) (evicted uint64, err error
 		return 0, err
 	}
 
-	if reserveSizeChange > 0 {
-		batch = new(leveldb.Batch)
-		if err := db.incReserveSizeInBatch(batch, -int64(reserveSizeChange)); err != nil {
+	// trigger garbage collection if we reached the capacity
+	if gcSize >= db.cacheCapacity {
+		db.triggerGarbageCollection()
+	}
+
+	return reserveSizeChange, nil
+}
+
+var unpinBatchSize = 10000
+
+func (db *DB) unpinBatchChunks(id []byte, bin uint8) (uint64, error) {
+	loggerV1 := db.logger.V(1).Register()
+	var (
+		batch                  = new(leveldb.Batch)
+		gcSizeChange           int64 // number to add or subtract from gcSize and reserveSize
+		reserveSizeChange      uint64
+		totalReserveSizeChange uint64
+	)
+	unpin := func(item shed.Item) (stop bool, err error) {
+		addr := swarm.NewAddress(item.Address)
+		c, err := db.setUnpin(batch, addr)
+		if err != nil {
+			if !errors.Is(err, leveldb.ErrNotFound) {
+				return false, fmt.Errorf("unpin: %w", err)
+			}
+			// this is possible when we are resyncing chain data after
+			// a dirty shutdown
+			loggerV1.Debug("unreserve set unpin chunk failed", "chunk", addr, "error", err)
+		} else {
+			// we need to do this because a user might pin a chunk on top of
+			// the reserve pinning. when we unpin due to an unreserve call, then
+			// we should logically deduct the chunk anyway from the reserve size
+			// otherwise the reserve size leaks, since c returned from setUnpin
+			// will be zero.
+			reserveSizeChange++
+		}
+
+		gcSizeChange += c
+		return false, nil
+	}
+
+	var startItem *shed.Item
+	for {
+		currentBatchSize := 0
+		more := false
```

Comment:
I think it would be more readable if you put the `more` variable as the loop param: `for more := fale; !more; {` and get rid of the break statement at the end of the loop.

Reply:
So if more is true we need to continue. And the initial value needs to be false. This is why I am not sure if it can be done using this type of for loop.

Reply:
Yes, you're right, I miss that.